File: Graph/python/dfs/stack.py
class Node:
    def __init__(self, val):  # Initialize a node with a value and a pointer to the next node
        self.elem = val
        self.next = None

class Stack:
    def __init__(self):  # Initialize an empty stack with a private top attribute
        self.__top = None

    def push(self, val):  # Push a new value onto the stack
        newNode = Node(val)
        newNode.next = self.__top  # New node points to the current top
        self.__top = newNode  # Update top to new node

    def pop(self):  # Remove and return the top element, return None if empty
        if self.__top is None:
            return None
        temp = self.__top.elem  # Store top element
        self.__top = self.__top.next  # Move top to next node
        return temp
    
    def peek(self):  # Return top element without removing it, return None if empty
        return None if self.__top is None else self.__top.elem
    
    def isEmpty(self):  # Check if stack is empty
        return self.__top is None

File: Graph/java/dfs/Stack.java
package dfs;

class Node {
    Integer elem; // Stores the element of the node
    Node next; // Reference to the next node

    Node(Integer val) {
        this.elem = val;
        this.next = null;
    }
}

public class Stack {
    private Node top; // Top of the stack, made private to enforce encapsulation

    // Push a new element onto the stack
    public void push(Integer val) {
        Node newNode = new Node(val);
        newNode.next = this.top;
        this.top = newNode;
    }

    // Returns the top element of the stack without removing it
    public Integer peek() {
        if (this.top == null) {
            return -1; // Indicating stack is empty
        }
        return this.top.elem;
    }

    // Removes and returns the top element of the stack
    public Integer pop() {
        if (this.top == null) {
            return null; // Stack is empty
        }
        Integer t1 = this.top.elem;
        this.top = this.top.next;
        return t1;
    }

    // Checks if the stack is empty
    public boolean isEmpty() {
        return this.top == null;
    }
}

